Math branch

mclark7-exe opened the pull request.

mclark7-exe committed:
Added function commonDenominator

Added helper function that will take in two mantissas and gives them a common denominator for easier addition and subtraction.

mclark7-exe committed:
Implemented commonDenominator

Implemented commonDenominator function and added a test of the function to main().

mclark7-exe committed:
Updated add and subtract functions

Added functionality to add and subtract functions. (Functions can't convert results into string yet)

mclark7-exe committed:
Added function numberOfDigits

Added helper function that returns the number of digits in an integer

mclark7-exe committed:
Added convertToCString function

Added function that takes in a characteristic and mantissa and converts it into a cstring, updated add and subtract functions to return cstrings

mclark7-exe committed:
Updated multiply and divide functions

Added functionality to multiply and divide functions

mclark7-exe committed:
Added removeInsignificantDigits function

Added helper function to remove insignificant digits from a cstring

mclark7-exe committed:
Negative numbers bugfix

Added functionality for negative numbers and greatly simplified subtract function

brandonc18 commented on main.cpp:
```diff
@@ -1,5 +1,6 @@
 
 #include <iostream>
+#include <string.h>
```
string.h is both unused and not allowed to be used for this project, so it can be removed.

brandonc18 replied:
This is both unused and not allowed for this assignment.

brandonc18 commented on main.cpp:
```diff
@@ -70,8 +77,36 @@ int main()
+    }
+    else
+    {
+        //display error message
```
Error comment seems unnecessary considering the message in the cout

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    n2 = abs(n2);
+    commonDenominator(n1, d1, n2, d2);
+
+    //handle negatives
```
You use this exact block of code in three different functions, consider putting it in a small helper function

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
-    result[8] = '6';
-    result[9] = '\0';
+    //check for invalid parameters (denominators should not be zero or negative)
+    if(d1 <=0 || d2 <=0)
```
You also reuse this block of code a few times, consider turning it into a small helper function.

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+        return false;
+    }
+
+    int lcm;
```
Not a great variable name

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    if (c1 < 0 && n1 >= 0) n1 *= -1;
+    if (c2 < 0 && n2 >= 0) n2 *= -1;
+
+    int improperNumerator1 = (c1 * d1) + n1;
```
Consider adding some comments to the math here

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    if (c1 < 0 && n1 >= 0) n1 *= -1;
+    if (c2 < 0 && n2 >= 0) n2 *= -1;
+
+    int improperNumerator1 = (c1 * d1) + n1;
```
Consider adding some comments to the math here as well

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    int improperNumerator1 = (c1 * d1) + n1;
+    int improperNumerator2 = (c2 * d2) + n2;
+    int resultNumerator = improperNumerator1 * d2;
+    int resultDenominator = d1 * improperNumerator2;
```
If resultDenominator is 0, you will end up in an infinite loop

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    {
+        resultCharacteristic++;
+        resultNumerator -= resultDenominator;
+    }
```
Doesn't handle cases with negative resultNumerator

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+    int resultDenominator = d1 * d2;
+    int resultCharacteristic = 0;
+
+    while (resultNumerator >= resultDenominator) {
```
Doesn't handle cases with negative resultNumerator

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+
+bool removeInsignificantDigits(char numCString [], int length)
+{
+    for (int i = length - 2; i >= 0; i--)
```
Maybe add a simple description comment

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
+        resultCString[0] = '-';
+        characteristic *= -1;
+    }
+
```
Could maybe make these three into an if else if chain

brandonc18 commented on main.cpp:
```diff
@@ -90,47 +125,273 @@ bool mantissa(const char numString[], int& numerator, int& denominator)
-    //hard coded return value to make the code compile
-    //you will have to come up with an algorithm to subtract the two numbers
-    return true;
+    if (add(c1, n1, d1, c2 * -1, n2, d2, result, len)) return true;
```
Maybe add comment explaining that the *-1 for easier understanding

mclark7-exe committed:
divide by 0 bugfix

fixed an error that caused an infinite loop when attempting to divide by zero, as well as other changes suggested by peer review

mclark7-exe committed:
Overflow bugfix

Added handling for overflow errors

mclark7-exe committed:
Merge branch 'main' into math_branch
